implementations/rust/ockam/ockam_api/src/cli_state/cli_state.rs

```diff
@@ -41,6 +41,32 @@ pub struct CliState {
     database: SqlxDatabase,
     application_database: SqlxDatabase,
     tracing_enabled: TracingEnabled,
+    tx: Option<tokio::sync::broadcast::Sender<ChannelMessageType>>,
+}
+
+const DEFAULT_CHANNEL_CAPACITY: usize = 16;
+pub type ChannelMessageType = String;
+pub const CHANNEL_POISON_PILL_MESSAGE: &str = "exit";
+pub const CHANNEL_MESSAGE_DISPLAY_DELAY_MS: u64 = 1000;
+
+// Broadcast channel support.
+impl CliState {
+    pub fn open_channel(&mut self) -> tokio::sync::broadcast::Sender<ChannelMessageType> {
```

Comment:
I would expect the channel to be opened for the life of the `CliState` and be closed only when `CliState` is dropped. Right now we have various places where the channel is closed, which is going to kill the channel. This is going to create issues if we call `get_named_identity_or_default` twice for example.

I would propose instead to:

 - Start the broadcast channel when `CliState` is started
 - Close it when `CliState` is dropped
 - When running a command, concurrently race that command + a small delay with a process that displays event as they arrive.
    - When the command + delay finished, we stop both computations
    - The spinner / display computation runs indefinitely, with no need for a poison pill. It just stops when the command has finished executing.

Reply:
Another thing I would keep in mind is having a way to support processing the messages in different ways:
1. Overwrite the spinner message
2. Close the previous spinner to persist the previous message and print the current one in a new spinner
3. Print the received message without any spinner
4. Ideally, we would also need another kind of message to print an error message and exit the program

With these kinds of messages we would cover most, if not all, the types of messages we show to the user.

Reply:
> I would expect the channel to be opened for the life of the `CliState` and be closed only when `CliState` is dropped. Right now we have various places where the channel is closed, which is going to kill the channel. This is going to create issues if we call `get_named_identity_or_default` twice for example.
> 
> I would propose instead to:
> 
> * Start the broadcast channel when `CliState` is started
> * Close it when `CliState` is dropped
> * When running a command, concurrently race that command + a small delay with a process that displays event as they arrive.
>   
>   * When the command + delay finished, we stop both computations
>   * The spinner / display computation runs indefinitely, with no need for a poison pill. It just stops when the command has finished executing.

I did as you suggested and it worked :)

Reply:
> Another thing I would keep in mind is having a way to support processing the messages in different ways:
> 
> 1. Overwrite the spinner message
> 2. Close the previous spinner to persist the previous message and print the current one in a new spinner
> 3. Print the received message without any spinner
> 4. Ideally, we would also need another kind of message to print an error message and exit the program
> 
> With these kinds of messages we would cover most, if not all, the types of messages we show to the user.

Great suggestion. We can move these concerns into another PR to generalize the progress reporting and applying it more broadly to all commands.